File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/data/AuthRepository.kt
package com.barkfluff.BarkCloud.data

import barkcloud.identity.IdentityApiOuterClass.AuthRequest
import barkcloud.identity.IdentityApiOuterClass.AuthResponse
import barkcloud.identity.IdentityApiOuterClass.LogoutRequest
import com.barkfluff.BarkCloud.grpc.AuthErrorCodes
import com.barkfluff.BarkCloud.grpc.GrpcManager
import com.barkfluff.BarkCloud.grpc.errorCode
import io.grpc.StatusRuntimeException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

sealed class AuthResult {
    data object Success : AuthResult()
    data object OtpRequired : AuthResult()
    data object InvalidCredentials : AuthResult()
    data class OtherError(val message: String) : AuthResult()
}

class AuthRepository(
    private val grpcManager: GrpcManager,
    private val globalParam: GlobalParam,
) {

    suspend fun auth(
        login: String,
        password: String,
        otpCode: String? = null,
    ): AuthResult = withContext(Dispatchers.IO) {
        val request = buildAuthRequest(login, password, otpCode)
        try {
            val response: AuthResponse = grpcManager.identityStub().auth(request)
            persist(response)
            AuthResult.Success
        } catch (e: StatusRuntimeException) {
            when (e.errorCode()) {
                AuthErrorCodes.OTP_REQUIRED -> AuthResult.OtpRequired
                AuthErrorCodes.INVALID_CREDENTIALS -> AuthResult.InvalidCredentials
                else -> AuthResult.OtherError(e.status.description ?: e.message ?: "gRPC error")
            }
        } catch (e: Exception) {
            AuthResult.OtherError(e.message ?: e::class.java.simpleName)
        }
    }

    /** Отзыв текущей сессии на сервере (best-effort: ошибки игнорируются). */
    suspend fun logout() = withContext(Dispatchers.IO) {
        runCatching { grpcManager.identityStub().logout(LogoutRequest.getDefaultInstance()) }
        Unit
    }

    private fun buildAuthRequest(
        login: String,
        password: String,
        otpCode: String?,
    ): AuthRequest {
        val builder = AuthRequest.newBuilder()
            .setPassword(password)
        if (login.contains('@')) {
            builder.setEmail(login)
        } else {
            builder.setUsername(login)
        }
        if (!otpCode.isNullOrBlank()) {
            builder.setOtpCode(otpCode)
        }
        return builder.build()
    }

    private fun persist(response: AuthResponse) {
        val access = response.accessToken
        val refresh = response.refreshToken
        globalParam.accessToken = access.value
        globalParam.accessTokenExpiresAt = access.expirationDate.seconds * 1000L
        globalParam.refreshToken = refresh.value
        globalParam.refreshTokenExpiresAt = refresh.expirationDate.seconds * 1000L
    }
}

File: Mac/BarkCloudKit/Sources/BarkCloudKit/Data/Auth/AuthRepository.swift
import Foundation
import GRPCCore
import SwiftProtobuf

public final class AuthRepository: Sendable {
    private let grpc: GrpcManager
    private let session: SessionStore

    public init(grpc: GrpcManager, session: SessionStore) {
        self.grpc = grpc
        self.session = session
    }

    public func auth(login: String, password: String, otpCode: String? = nil) async -> AuthResult {
        do {
            let stub = try await grpc.identityStub()
            var req = Barkcloud_Identity_AuthRequest()
            req.password = password
            if login.contains("@") {
                req.email = login
            } else {
                req.username = login
            }
            if let otp = otpCode, !otp.isEmpty {
                req.otpCode = otp
            }
            let response = try await stub.auth(req)
            await persist(response)
            return .success
        } catch let err as RPCError {
            switch err.errorCode {
            case AuthErrorCodes.otpRequired:
                return .otpRequired
            case AuthErrorCodes.invalidCredentials:
                return .invalidCredentials
            default:
                return .otherError(err.message)
            }
        } catch {
            return .otherError(error.localizedDescription)
        }
    }

    /// Серверный отзыв текущей сессии. Использует auth-контекст из интерсепторов
    /// (`x-auth-token` + устройство), поэтому вызывается до очистки токенов.
    /// Best-effort: ошибки игнорируются — сессия могла истечь или отсутствует сеть.
    public func logout() async {
        do {
            let stub = try await grpc.identityStub()
            _ = try await stub.logout(Barkcloud_Identity_LogoutRequest())
        } catch {
            // best-effort: локальная очистка всё равно выполнится
        }
    }

    @MainActor
    private func persist(_ response: Barkcloud_Identity_AuthResponse) {
        let access = response.accessToken
        let refresh = response.refreshToken
        session.accessToken = access.value
        session.accessTokenExpiresAt = access.hasExpirationDate ? Self.date(from: access.expirationDate) : nil
        session.refreshToken = refresh.value
        session.refreshTokenExpiresAt = refresh.hasExpirationDate ? Self.date(from: refresh.expirationDate) : nil
        session.sessionExpired = false
    }

    private static func date(from ts: SwiftProtobuf.Google_Protobuf_Timestamp) -> Date {
        Date(timeIntervalSince1970: TimeInterval(ts.seconds) + TimeInterval(ts.nanos) / 1_000_000_000)
    }
}
